skmonitor/extern/native.c

```diff
@@ -0,0 +1,58 @@
+#include <ctype.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <time.h>
+#include <unistd.h>
+
+void* SKIP_SKMonitor_createTime(__uint64_t s, __uint64_t ns);
+void sk_global_lock();
+void sk_global_unlock();
+char* sk2c_string(char*);
+
+void* SKIP_SKMonitor_now() {
+  struct timespec now;
+  timespec_get(&now, TIME_UTC);
+  return SKIP_SKMonitor_createTime(now.tv_sec, now.tv_nsec);
+}
+
+__uint8_t SKIP_SKMonitor_checkFifo(char* skfifo) {
+  char* fifo = sk2c_string(skfifo);
+  if (access(fifo, F_OK) == 0) {
+    if (fifo != skfifo) free(fifo);
+    return 1;
+  }
+  if (fifo != skfifo) free(fifo);
+  return 0;
+}
+
+__uint8_t SKIP_SKMonitor_appendToFifo(char* skfifo, char* sktrace) {
+  char* trace = sk2c_string(sktrace);
+  char* fifo = sk2c_string(skfifo);
+  int fifo_write = open(fifo, O_WRONLY | O_APPEND | O_NONBLOCK);
+  if (fifo_write < 0) {
+    return 0;
+  }
+  char* b = trace;
+  const char* s;
+  for (s = trace; *s; ++s)
+    ;
+  size_t n = (s - trace);
+  size_t r = 0;
+  sk_global_lock();
```

Comment:
Why do we need to do this? Do different processes write to the same fifo?

Reply:
In order to avoid creating a fifo management mechanism, it's true that in the first version, concurrent processes will probably write to the same fifo.

Reply:
I'm ok with this for first version, but I feel strongly that we should prioritise avoiding the lock. Logging can be very high throughput and this would then introduce a major scaling bottleneck.